app/models/supporter.rb

```diff
@@ -147,22 +206,30 @@ def filled_address_fields?
   end
 
   def update_primary_address
-    if self.changes.slice(*ADDRESS_FIELDS).any? #changed an address field
-      if filled_address_fields?
-        if primary_address.nil?
-          self.addresses.build(address_field_attributes)
-        else
-          primary_address.update(address_field_attributes)
-        end
-      elsif primary_address.present?
-        prim_addr = primary_address
-        self.update(primary_address: nil)
-        self.addresses.delete(prim_addr)
-        prim_addr.destroy
+    if filled_address_fields?
+      if primary_address.nil?
+        self.addresses.build(address_field_attributes)
+      else
+        primary_address.update(address_field_attributes)
       end
+    elsif primary_address.present? && empty_new_address_fields?
+      delete_primary_address
     end
   end
 
+  def empty_new_address_fields?
+    address.blank? && city.blank? && state_code.blank? && country.blank? && zip_code.blank?
+  end
+
+  def delete_primary_address
+    self.primary_address.update(supporter_id: nil)
+    prim_addr = self.primary_address
+    self.addresses.delete(prim_addr)
+    address_field_attributes.each { |address_attribute, _| self[address_attribute] = nil }
+    self.update_attributes(primary_address: nil)
+    prim_addr.destroy
```

Comment:
This seems like a lot of different things going on for a delete. What was the reasoning for each of these lines?

Reply:
Because I couldn't just destroy the primary address because of the relationships between the supporter and the address.

I started by removing the reference from the primary address to the supporter, then I deleted the primary address from the list of supporter addresses, then I make the primary address from the supporter nil and finally destroy the primary address because there's no supporter referencing it and it is not referencing a supporter anymore.

Reply:
I'm going to remove line 228 because it's not needed.

Reply:
What part of the relationship was prevented this from working more simply though?

Reply:
If I simply do a `self.primary_address.destroy`, I get this error:

```
ActiveRecord::InvalidForeignKey:
       PG::ForeignKeyViolation: ERROR:  update or delete on table "supporter_addresses" violates foreign key constraint "fk_rails_1007f7f722" on table "supporters"
       DETAIL:  Key (id)=(2) is still referenced from table "supporters".
       : DELETE FROM "supporter_addresses" WHERE "supporter_addresses"."id" = $1
```

This `fk_rails_1007f7f722` is `primary_address_id`.

If I run:

```ruby
    self.primary_address.update(supporter_id: nil)
    prim_addr = self.primary_address
    self.update(primary_address: nil)
    prim_addr.destroy
 ```
 
It works for deleting the primary address from the supporter but the supporter continues to have it as a supporter_address, that's why I also delete it from the supporter addresses.

We discussed something similar on [this PR](https://github.com/CommitChange/houdini/pull/140).

Reply:
I think we really need to scrap that foreign key constraint. Let's use ActiveRecord validation instead.

Reply:
Did that [here](https://github.com/CommitChange/houdini/pull/243).